app/upgrades/mainnet/v3/upgrade.go

```diff
@@ -17,20 +22,94 @@ import (
 const UpgradeName = "v3"
 
 func UpgradeHandler(
-	mm *module.Manager, configurator module.Configurator, marketmakerkeeper marketmakerkeeper.Keeper, liquiditykeeper liquiditykeeper.Keeper) upgradetypes.UpgradeHandler {
+	mm *module.Manager, configurator module.Configurator, marketmakerKeeper marketmakerkeeper.Keeper,
+	liquidityKeeper liquiditykeeper.Keeper, farmKeeper farmkeeper.Keeper, farmingKeeper farmingkeeper.Keeper) upgradetypes.UpgradeHandler {
 	return func(ctx sdk.Context, plan upgradetypes.Plan, vm module.VersionMap) (module.VersionMap, error) {
 		newVM, err := mm.RunMigrations(ctx, configurator, vm)
 		if err != nil {
 			return newVM, err
 		}
 
 		// Set newly added liquidity param
-		liquiditykeeper.SetMaxNumMarketMakingOrderTicks(ctx, liquiditytypes.DefaultMaxNumMarketMakingOrderTicks)
+		liquidityKeeper.SetMaxNumMarketMakingOrderTicks(ctx, liquiditytypes.DefaultMaxNumMarketMakingOrderTicks)
 
 		// Set param for new market maker module
-		marketmakerparams := marketmakertypes.DefaultParams()
-		marketmakerparams.DepositAmount = sdk.NewCoins(sdk.NewCoin("ucre", sdk.NewInt(1000000000)))
-		marketmakerkeeper.SetParams(ctx, marketmakerparams)
+		marketmakerParams := marketmakertypes.DefaultParams()
+		marketmakerParams.DepositAmount = sdk.NewCoins(sdk.NewCoin("ucre", sdk.NewInt(1000000000)))
+		marketmakerKeeper.SetParams(ctx, marketmakerParams)
+
+		// Unstake all staked coins from x/farming and start farming on x/farm.
+		stakedCoinsByFarmer := map[string]sdk.Coins{}
+		var farmerAddrs []sdk.AccAddress
+		farmingKeeper.IterateStakings(
+			ctx, func(denom string, farmerAddr sdk.AccAddress, staking farmingtypes.Staking) (stop bool) {
+				farmer := farmerAddr.String()
+				if _, ok := stakedCoinsByFarmer[farmer]; !ok {
+					farmerAddrs = append(farmerAddrs, farmerAddr)
+				}
+				stakedCoinsByFarmer[farmer] = stakedCoinsByFarmer[farmer].
+					Add(sdk.NewCoin(denom, staking.Amount))
+				return false
+			})
+		farmingKeeper.IterateQueuedStakings(
+			ctx, func(_ time.Time, denom string, farmerAddr sdk.AccAddress, queuedStaking farmingtypes.QueuedStaking) (stop bool) {
+				farmer := farmerAddr.String()
+				if _, ok := stakedCoinsByFarmer[farmer]; !ok {
+					farmerAddrs = append(farmerAddrs, farmerAddr)
+				}
+				stakedCoinsByFarmer[farmer] = stakedCoinsByFarmer[farmer].
+					Add(sdk.NewCoin(denom, queuedStaking.Amount))
+				return false
+			})
+		for _, farmerAddr := range farmerAddrs {
+			if err := farmingKeeper.Unstake(ctx, farmerAddr, stakedCoinsByFarmer[farmerAddr.String()]); err != nil {
+				return nil, err
+			}
+			for _, stakedCoin := range stakedCoinsByFarmer[farmerAddr.String()] {
+				if _, err := farmKeeper.Farm(ctx, farmerAddr, stakedCoin); err != nil {
+					return nil, err
+				}
+			}
+		}
```

Comment:
Since the `farming` module allocates farming rewards for epoch basis, am I understanding correctly that we don't need to take rewards into consideration in this upgrade handler?